### PR TITLE
FnCAS Tweaks

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: FnCAS/fncas/mathutil.h [hunk 1]
```diff
@@ -135,8 +153,9 @@ inline std::vector<double> Backtracking(F&& eval_function,
       break;
     }
   }
+  OptimizerLogger().Log(std::string("Backtracking: Final value = ") + current::ToString(test_f_value));
 
-  return test_point;
+  return std::make_pair(test_f_value, test_point);
```
Comment (mzhurovich) [inline FnCAS/fncas/mathutil.h]: `std::move(test_point)`?

Comment (mzhurovich) [inline FnCAS/fncas/mathutil.h]: A couple of thoughts:
`OptimizationResult` has the same structure. We can unify them under some `PointAndValue` structure.
We also use `std::vector<double>` explicitly to define a point in R<sup>n</sup>. As we go further with more sophisticated linear algebra methods, a lack of proper mathematical vector/matrix notation could become troublesome. I don't suggest to spend time on linear algebra engine now, but maybe it's better to add some `Point` class (aliased to `std::vector<double>`) abstraction at this stage, so we can easily upgrade math engine later. What fo you think?

File: FnCAS/fncas/optimize.h [hunk 2]
```diff
@@ -117,45 +121,86 @@ class Optimizer : noncopyable {
...
 
   OptimizationResult Optimize(const std::vector<double>& starting_point) const override {
-    size_t max_steps = 5000;   // Maximum number of optimization steps.
-    double step_factor = 1.0;  // Gradient is multiplied by this factor.
+    size_t max_steps = 5000;                           // Maximum number of optimization steps.
+    double step_factor = 1.0;                          // Gradient is multiplied by this factor.
+    double min_absolute_per_step_improvement = 1e-25;  // Terminate early if the absolute improvement is less than this.
+    double min_relative_per_step_improvement = 1e-25;  // Terminate early if the relative improvement is less than this.
+    double no_improvement_steps_to_terminate =
+        2;  // Terminate early if the improvement condition is broken for this number of iterations.
```
Comment (mzhurovich) [inline FnCAS/fncas/optimize.h]: clang-format prefer no line wraps in comments?

File: FnCAS/test.cc [hunk 3]
```diff
@@ -286,39 +286,40 @@ TEST(FNCAS, OptimizationOfAMemberFunctionForwardsParameters) {
...
   const auto result = fncas::ConjugateGradientOptimizer<PolynomialFunction>().Optimize({5.0, 20.0});
   EXPECT_NEAR(0.0, result.value, 1e-6);
   ASSERT_EQ(2u, result.point.size());
   EXPECT_NEAR(0.0, result.point[0], 1e-6);
   EXPECT_NEAR(0.0, result.point[1], 1e-6);
 }
 
-TEST(FNCAS, OptimizationOfRosenbrockUsingConjugateGradient) {
+TEST(FnCAS, OptimizationOfRosenbrockUsingConjugateGradient) {
+  // fncas::ScopedLogToStderr scope;
```
Comment (mzhurovich) [inline FnCAS/test.cc]: Debug leftovers?

**Comment by mzhurovich** [comment]
@dkorolev I've found a bug causing segfault. 
`optimize.h` line 111:
`const Optional<OptimizerParameters> Parameters() const { return parameters_; }`
should be `const Optional<OptimizerParameters>&`

File: FnCAS/fncas/optimize.h [hunk 4]
```diff
@@ -117,45 +121,86 @@ class Optimizer : noncopyable {
...
+          OptimizerLogger().Log("GradientDescentOptimizer: Value " + current::ToString(value) + " at step " +
+                                current::ToString(step));
+          best_candidate = std::min(best_candidate, std::make_pair(value, candidate_point));
+        }
+      }
+      if (!has_valid_candidate) {
+        CURRENT_THROW(FnCASOptimizationException("!fncas::IsNormal(value)"));
+      }
+      if (best_candidate.first / current.first > 1.0 - min_relative_per_step_improvement ||
+          current.first - best_candidate.first < min_absolute_per_step_improvement) {
```
Comment (mzhurovich) [inline FnCAS/fncas/optimize.h]: `std::abs(...)` for both statements?

File: FnCAS/fncas/mathutil.h (hunk 1 above)
Comment (dkorolev) [inline FnCAS/fncas/mathutil.h]: :+1: 5ad49be

File: FnCAS/test.cc (hunk 3 above)
Comment (dkorolev) [inline FnCAS/test.cc]: Sorry about that, fixes.

File: FnCAS/fncas/optimize.h (hunk 4 above)
Comment (dkorolev) [inline FnCAS/fncas/optimize.h]: The objective function should be strictly decreasing, shouldn't it?

**Comment by dkorolev** [comment]
@mzhurovich Thanks for the catch (the issue is not the return type though but assigning it to a const reference -- fixed regardless) and for the comments. Fixed everything, PTAL.

**Comment by mzhurovich** [comment]
@dkorolev :+1:

**Pull request merged**